keanu-project/src/main/java/io/improbable/keanu/network/ProtobufReader.java

```diff
@@ -0,0 +1,236 @@
+import io.improbable.keanu.tensor.intgr.IntegerTensor;
+import io.improbable.keanu.vertices.LoadParentVertex;
+import io.improbable.keanu.vertices.LoadVertexValue;
+import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.VertexLabel;
+import io.improbable.keanu.vertices.bool.BoolVertex;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.intgr.IntegerVertex;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Parameter;
+import java.util.HashMap;
+import java.util.Map;
+
+public class ProtobufReader implements NetworkReader {
+
+    private final Map<Vertex, KeanuSavedBayesNet.VertexValue> savedValues;
+
+    public ProtobufReader() {
+        savedValues = new HashMap<>();
+    }
+
+    @Override
+    public void loadValue(Vertex vertex) {
+        throw new IllegalArgumentException("Cannot Load value for Untyped Vertex");
+    }
+
+    public BayesianNetwork loadNetwork(InputStream input) throws IOException {
+        Map<KeanuSavedBayesNet.VertexID, Vertex> instantiatedVertices = new HashMap<>();
+        KeanuSavedBayesNet.BayesianNetwork parsedNet = KeanuSavedBayesNet.BayesianNetwork.parseFrom(input);
+
+        for (KeanuSavedBayesNet.Vertex vertex : parsedNet.getVerticesList()) {
+            Vertex newVertex = createVertexFromProtoBuf(vertex, instantiatedVertices);
+            instantiatedVertices.put(vertex.getId(), newVertex);
+        }
+
+        BayesianNetwork bayesNet = new BayesianNetwork(instantiatedVertices.values());
+
+        loadDefaultValues(parsedNet, instantiatedVertices, bayesNet);
+
+        return bayesNet;
+    }
+
+    @Override
+    public void loadValue(DoubleVertex vertex) {
+        KeanuSavedBayesNet.VertexValue value = savedValues.get(vertex);
+        DoubleTensor tensor = extractDoubleValue(value);
+        vertex.setValue(tensor);
+    }
+
+    private DoubleTensor extractDoubleValue(KeanuSavedBayesNet.VertexValue value) {
+        if (value.getValueTypeCase() != KeanuSavedBayesNet.VertexValue.ValueTypeCase.DOUBLEVAL) {
+            throw new IllegalArgumentException("Non Double Value specified for Double Vertex");
+        } else {
+            return DoubleTensor.create(
+                Doubles.toArray(value.getDoubleVal().getValuesList()),
+                Longs.toArray(value.getDoubleVal().getShapeList()));
+        }
+    }
+
+    private void loadDefaultValues(KeanuSavedBayesNet.BayesianNetwork parsedNet,
+                                   Map<KeanuSavedBayesNet.VertexID, Vertex> instantiatedVertices,
+                                   BayesianNetwork bayesNet) {
+        for (KeanuSavedBayesNet.StoredValue value : parsedNet.getDefaultStateList()) {
+            Vertex targetVertex = null;
+
+            if (value.hasId()) {
+                targetVertex = instantiatedVertices.get(value.getId());
+            }
+
+            if (value.getVertexLabel() != "") {
```

Comment:
Either `!value.getVertexLabel().isEmpty()` or `value.getVertexLabel().equals("")`. Generally avoid != unless you're actually going for object equality. In Java's case Strings are interned, which is why this usually works.

Reply:
Done